quantum/process_keycode/process_joystick.c

```diff
@@ -0,0 +1,160 @@
+#include "process_joystick.h"
+
+#include <quantum/joystick.h>
+#include <quantum/quantum_keycodes.h>
+#include <quantum/config_common.h>
+
+#ifdef __AVR__
+# include <drivers/avr/analog.h>
+#endif
+
+#include <string.h>
+#include <math.h>
+
+bool process_joystick_buttons(uint16_t keycode, keyrecord_t *record);
+
+bool process_joystick(uint16_t keycode, keyrecord_t *record){
+  
+  if (process_joystick_buttons(keycode, record) 
+    && (joystick_status.status & JS_UPDATED)>0){
+    send_joystick_packet(&joystick_status);
+    joystick_status.status &= ~JS_UPDATED;
+  }
+  
+  return true;
+}
+
+__attribute__ ((weak))
+void joystick_task(void){
+  if (process_joystick_analogread() && (joystick_status.status & JS_UPDATED)){
+    send_joystick_packet(&joystick_status);
+    joystick_status.status &= ~JS_UPDATED;
+  }
+}
+
+
+
+bool process_joystick_buttons(uint16_t keycode, keyrecord_t *record){
+  
+  if (keycode < JS_BUTTON0 || keycode > JS_BUTTON_MAX){
+    return true;
+  } else {
+    if (record->event.pressed){
+      joystick_status.buttons[(keycode-JS_BUTTON0)/8] |= 1<<(keycode%8);
+    } else {
+      joystick_status.buttons[(keycode-JS_BUTTON0)/8] &= ~(1<<(keycode%8));
+    }
+    
+    joystick_status.status |= JS_UPDATED;
+  }
+  
+  return true;
+}
+
+uint8_t savePinState(uint8_t pin){
+#ifdef __AVR__
+  uint8_t pinNumber = pin & 0xF;
+  return ((PORTx_ADDRESS(pin) >> pinNumber) & 0x1) << 1 
+       | ((DDRx_ADDRESS(pin) >> pinNumber) & 0x1) ;
+#else
+  return 0;
+#endif
+}
+
+void restorePinState(uint8_t pin, uint8_t restoreState){
+#ifdef __AVR__
+  uint8_t pinNumber = pin & 0xF;
+  PORTx_ADDRESS(pin) = (PORTx_ADDRESS(pin) & ~_BV(pinNumber)) | (((restoreState >> 1) & 0x1) << pinNumber);
+  DDRx_ADDRESS(pin) = (DDRx_ADDRESS(pin) & ~_BV(pinNumber)) | ((restoreState & 0x1) << pinNumber);
+#else
+  return;
+#endif
+}
```

Comment:
Why not just use `readPin()`, `writePin*()`?

Reply:
readPin reads the state of input pins and writePin writes output pins, whereas here we need to read the state of output pins and data direction to be able to restore later.